eth/db/journal.py

```diff
@@ -24,237 +26,258 @@ class DeletedEntry:
         """
-        return first(self.journal_data.keys())
+        return first(self._journal_data.keys())
 
     @property
     def is_flattened(self) -> bool:
         """
         :return: whether there are any explicitly committed checkpoints
         """
-        return len(self.journal_data) < 2
+        return len(self._checkpoint_stack) < 2
 
     @property
-    def latest_id(self) -> uuid.UUID:
+    def last_changeset_index(self) -> JournalDBCheckpoint:
         """
         Returns the id of the latest changeset
         """
         # last() was iterating through all values, so first(reversed()) gives a 12.5x speedup
-        return first(reversed(self.journal_data.keys()))
-
-    @property
-    def latest(self) -> Dict[bytes, Union[bytes, DeletedEntry]]:
-        """
-        Returns the dictionary of db keys and values for the latest changeset.
-        """
-        return self.journal_data[self.latest_id]
-
-    @latest.setter
-    def latest(self, value: Dict[bytes, Union[bytes, DeletedEntry]]) -> None:
-        """
-        Setter for updating the *latest* changeset.
-        """
-        self.journal_data[self.latest_id] = value
-
-    def is_empty(self) -> bool:
-        return len(self.journal_data) == 0
+        # Interestingly, an attempt to cache this value caused a slowdown.
+        return first(reversed(self._journal_data.keys()))
 
-    def has_changeset(self, changeset_id: uuid.UUID) -> bool:
-        return changeset_id in self.journal_data
+    def has_changeset(self, changeset_id: JournalDBCheckpoint) -> bool:
+        # another option would be to enforce monotonically-increasing changeset ids, so we can do:
+        # checkpoint_idx = bisect_left(self._checkpoint_stack, changeset_id)
+        # (then validate against length and value at index)
+        return changeset_id in self._checkpoint_stack
 
-    def record_changeset(self, custom_changeset_id: uuid.UUID = None) -> uuid.UUID:
+    def record_changeset(
+            self,
+            custom_changeset_id: JournalDBCheckpoint = None) -> JournalDBCheckpoint:
         """
-        Creates a new changeset. Changesets are referenced by a random uuid4
+        Creates a new changeset. Changesets are referenced by a random int
         to prevent collisions between multiple changesets.
         """
         if custom_changeset_id is not None:
-            if custom_changeset_id in self.journal_data:
+            if custom_changeset_id in self._journal_data:
                 raise ValidationError(
                     "Tried to record with an existing changeset id: %r" % custom_changeset_id
                 )
             else:
                 changeset_id = custom_changeset_id
         else:
-            changeset_id = uuid.uuid4()
+            changeset_id = get_next_checkpoint_id()
 
-        self.journal_data[changeset_id] = {}
+        self._journal_data[changeset_id] = {}
+        self._checkpoint_stack.append(changeset_id)
```

Comment:
Just to back my previous comment, here we are appending a `changeset_id` to the `checkpoint_stack`